### PR TITLE
ref(*): use S3 client instead of mc CLI

### PR DESCRIPTION
**Pull request by aledbf**
replaces #106

### CODE REVIEW

File: glide.yaml [hunk 1]
```diff
@@ -21,14 +25,15 @@ import:
...
-  - appengine
-  - appengine/memcache
-  - appengine/datastore
-  - appengine/user
+- package: github.com/aws/aws-sdk-go
+  version: 87b1e60a50b09e4812dee560b33a238f67305804
+  subpackages:
+  - aws
+  - aws/session
+  - service/s3
```
Comment (arschles) [inline glide.yaml]: nitpick: can you add a newline here?

File: pkg/gitreceive/build.go [hunk 2]
```diff
@@ -138,23 +125,22 @@ func build(conf *Config, builderKey, gitSha string) error {
...
-	if creds == nil {
-		// both key and secret are missing, proceed with no credentials
-		if usingDockerfile {
-			srcManifest = "/etc/deis-dockerbuilder-no-creds.yaml"
-		} else {
-			srcManifest = "/etc/deis-slugbuilder-no-creds.yaml"
-		}
-	} else if err == nil {
+
+	creds := storage.IsCredentialsOK()
```
Comment (arschles) [inline pkg/gitreceive/build.go]: Can you shorten this to `CredsOK`?

File: pkg/gitreceive/storage/auth.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+	accessSecretKeyFile = "/var/run/secrets/object/store/access-secret-key"
+)
+
+var (
+	errMissingKey    = fmt.Errorf("missing %s", accessKeyIDFile)
+	errMissingSecret = fmt.Errorf("missing %s", accessSecretKeyFile)
+	emptyAuth        = credentials.AnonymousCredentials
+)
+
+// getCreds gets storage credentials from accessKeyIDFile and accessSecretKeyFile.
```
Comment (arschles) [inline pkg/gitreceive/storage/auth.go]: nitpick: can you change this to read `getAuth gets...`?

File: pkg/gitreceive/storage/bucket.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+var (
+	ACLPublicRead = aws.String("public-read")
+)
+
+func BucketExists(svc *s3.S3, bucketName string) (bool, error) {
+	_, err := svc.HeadBucket(&s3.HeadBucketInput{
+		Bucket: aws.String(bucketName),
+	})
+	if err != nil {
```
Comment (arschles) [inline pkg/gitreceive/storage/bucket.go]: Is there a specific error that the AWS SDK returns when the API returns `404 Not Found`? If so, it would be great to only return `false, nil` in that case, and `false, err` in all other error cases.

**Comment by arschles** [comment]
@aledbf thanks for this PR - looks very good. I left a few small comments, so let me know when they're done and I'll do another test & review right away.

**Comment by aledbf** [comment]
@arschles ready for another review

**Comment by arschles** [comment]
ok, looking

File: pkg/gitreceive/storage/bucket.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+		if awsErr, ok := err.(awserr.Error); ok {
```
Comment (arschles) [inline pkg/gitreceive/storage/bucket.go]: :+1:

**Comment by arschles** [comment]
@aledbf I am testing now

**Comment by arschles** [comment]
@aledbf the build fails with the following:

```console
ENG000656:builder aaronschlesinger$ make bootstrap
docker run --rm -e GO15VENDOREXPERIMENT=1 -v /Users/aaronschlesinger/gocode/src/github.com/deis/builder:/go/src/github.com/deis/builder -w /go/src/github.com/deis/builder quay.io/deis/go-dev:0.3.0 glide install
Oops! Lock file does not match YAML configuration. Consider running 'update'
make: *** [bootstrap] Error 1
```

A `make glideup` should fix.

**Comment by aledbf** [comment]
@arschles done

**Comment by arschles** [comment]
@aledbf thanks - I'm still waiting for a build :(

**Comment by arschles** [comment]
@aledbf tested and everything works as expected, thanks!

one minor note - can you check for the `409

**Comment by aledbf** [comment]
@arschles done

**Comment by smothiki** [comment]
Please merge this after the PR https://github.com/deis/builder/pull/129 is merged

**Comment by bacongobbler** [comment]
@smothiki any particular reason why? merge conflicts? Personally I'm in favour of merging the bigger PRs first so there's less to change in those cases.

**Comment by arschles** [comment]
I'd like to get this reviewed and merged ASAP as it is working and will enable @aledbf to move onto his implementation of #115

**Comment by smothiki** [comment]
PR #129  has very few changes and touches only build.go. I thought merging conflicts after 129 gets merged is pretty easy than merging conflicts of this PR in 129 .
I'm okay if this is ready to merge. But PR 129 is an individual PR by it self which can get quick reviews

**Comment by arschles** [comment]
Per offline discussion between me and @smothiki I'm merging this now and marking #129 as `needs rebase`
